src/main/java/org/snomed/snowstorm/core/data/services/ConceptChangeHelper.java

```diff
@@ -255,128 +255,153 @@ private NativeSearchQueryBuilder newSearchQuery(BoolQueryBuilder branchUpdatesCr
-                Map.Entry<Long, Date> entrySet = iterator.next();
-                Long conceptId = entrySet.getKey();
-                Date endDate = entrySet.getValue();
+        Set<String> conceptsDeleted = new HashSet<>(conceptsEndedOnBranch);
 
-                try (final SearchHitsIterator<Concept> stream = elasticsearchTemplate.searchForStream(new NativeSearchQueryBuilder()
-                        .withQuery(
-                                boolQuery()
-                                        .must(termQuery(Description.Fields.CONCEPT_ID, conceptId.toString()))
-                                        .must(termQuery(Description.Fields.PATH, parentBranchPath))
-                                        .must(termQuery(Description.Fields.START, endDate.getTime()))
-                        )
-                        .withFields(Concept.Fields.CONCEPT_ID)
-                        .build(), Concept.class)) {
-                    stream.forEachRemaining(hit -> iterator.remove());
-                }
-            }
-        }
-
-        conceptsDeletedOnBranch.addAll(conceptsEnded.keySet());
-
-        // Example context: Task.
-        // Technique: If a Concept has been flagged as being replaced but there is no corresponding replacement, the Concept
-        // has in fact been deleted on the task.
-        Set<Long> conceptsNotReplaced = new HashSet<>();
-        Set<String> versionsReplaced = branchService.findLatest(branchPath).getVersionsReplaced(Concept.class);
-        if (!versionsReplaced.isEmpty()) {
-            Set<Long> conceptIds = getConceptsByInternalIdentifierAndPath(versionsReplaced);
+        // Case 2
+        if (!conceptsReplacedOnBranch.isEmpty()) {
             try (final SearchHitsIterator<Concept> stream = elasticsearchTemplate.searchForStream(new NativeSearchQueryBuilder()
                     .withQuery(
                             boolQuery()
-                                    .must(termsQuery(Concept.Fields.CONCEPT_ID, conceptIds))
+                                    .must(termsQuery(Concept.Fields.CONCEPT_ID, conceptsReplacedOnBranch))
                                     .must(termQuery(Concept.Fields.PATH, branchPath))
                                     .mustNot(existsQuery(Concept.Fields.END))
                     )
+                    .withFields(Concept.Fields.CONCEPT_ID)
                     .build(), Concept.class)) {
-                stream.forEachRemaining(hit -> {
-                    String id = hit.getContent().getConceptId();
-                    // Concept replacement found, therefore not deleted.
-                    if (id != null) {
-                        conceptIds.removeIf(e -> e.equals(Long.parseLong(id)));
-                    }
-                });
+                stream.forEachRemaining(hit -> conceptsReplacedOnBranch.removeIf(id -> id.equals(hit.getContent().getConceptId())));
             }
+        }
+
+        conceptsDeleted.addAll(conceptsReplacedOnBranch);
+
```

Comment:
Yes this looks much better now and has implemented the logic as discussed. Can you add a short description for case 1 and case 2 in the comment or java doc.